fv3core/fv3core/utils/functional_validation.py

```diff
@@ -0,0 +1,39 @@
+import copy
+from typing import Callable, Sequence, Tuple
+
+import numpy as np
+
+from pace.dsl.stencil import GridIndexing
+
+
+def get_subset_func(
+    grid_indexing: GridIndexing,
+    dims: Sequence[str],
+    n_halo: Tuple[Tuple[int, int], Tuple[int, int]] = ((0, 0), (0, 0)),
+) -> Callable[[np.ndarray], np.ndarray]:
+    origin, domain = grid_indexing.get_origin_domain(dims)
+    i_start = origin[0] - n_halo[0][0]
+    i_end = origin[0] + domain[0] + n_halo[0][1]
+    j_start = origin[1] - n_halo[1][0]
+    j_end = origin[1] + domain[1] + n_halo[1][1]
+
+    def subset(data):
+        return data[i_start:i_end, j_start:j_end, :]
+
+    return subset
+
+
+def get_set_nan_func(
+    grid_indexing: GridIndexing,
+    dims: Sequence[str],
+    n_halo: Tuple[Tuple[int, int], Tuple[int, int]] = ((0, 0), (0, 0)),
+) -> Callable[[np.ndarray], np.ndarray]:
+    subset = get_subset_func(grid_indexing=grid_indexing, dims=dims, n_halo=n_halo)
+
+    def set_nans(data):
+        safe = copy.deepcopy(data)
+        data[:] = np.nan
+        data_subset = subset(data)
+        data_subset[:] = subset(safe)
```

Comment:
I don't fully understand this function. Why do you need `data_subset` here if it's neither an argument or return?

Reply:
Data_subset is a view into `data`. By modifying `data_subset`, we modify `data`. I'll add this as a comment.
```suggestion
    def set_nans(data):
        safe = copy.deepcopy(data)
        data[:] = np.nan
        # data_subset is a view of data, so modifying data_subset modifies data
        data_subset = subset(data)
        data_subset[:] = subset(safe)
```